### PR TITLE
feat(planner): Support `TableFunction` in new planner

### LINKED ISSUE #5134: Support `numbers` system table in new planner
`numbers` is a primitive system table function. The reference is: https://databend.rs/doc/reference/sql/system-tables/system-numbers

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
Support `TableFunction`s in new planner, e.g. `numbers`.

## Changelog

- New Feature

## Related Issues

Fixes #5134

### CODE REVIEW

File: common/ast/src/parser/query.rs [hunk 1]
```diff
@@ -158,7 +158,7 @@ pub fn set_returning_function(i: Input) -> IResult<TableReference> {
         rule! {
             #ident ~ "(" ~ #comma_separated_list0(expr) ~ ")" ~ #table_alias?
         },
-        |(name, _, params, _, alias)| TableReference::SetReturningFunction {
+        |(name, _, params, _, alias)| TableReference::TableFunction {
```
Comment (andylokandy) [inline common/ast/src/parser/query.rs:161]: Remember to also rename L156

Comment (leiysky) [reply common/ast/src/parser/query.rs:161]: Done.

File: query/src/sql/planner/binder/select.rs [hunk 2]
```diff
@@ -31,23 +31,35 @@ use crate::sql::planner::binder::Binder;
...
 use crate::storages::Table;
+use crate::storages::ToReadDataSourcePlan;
+use crate::table_functions::TableFunction;
 
 impl Binder {
     #[async_recursion]
-    pub(super) async fn bind_query(&mut self, query: &Query) -> Result<BindContext> {
+    pub(super) async fn bind_query(
+        &mut self,
+        query: &Query,
```
Comment (BohuTANG) [inline query/src/sql/planner/binder/select.rs:45]: Is it better to move the 'bind_context' as the first function parameter?

**Review by BohuTANG — APPROVED** [review]

File: query/src/sql/planner/binder/select.rs (hunk 2 above)
Comment (leiysky) [reply query/src/sql/planner/binder/select.rs:45]: Got. I'll fix this later.

**Pull request merged**